x/btcstaking/keeper/power_dist_change.go

```diff
@@ -0,0 +1,172 @@
+package keeper
+
+import (
+	"context"
+
+	"cosmossdk.io/store/prefix"
+	"github.com/babylonchain/babylon/x/btcstaking/types"
+	"github.com/cosmos/cosmos-sdk/runtime"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+/* power distribution update */
+
+// UpdatePowerDist updates the voting power distribution of finality providers
+// and their BTC delegations
+func (k Keeper) UpdatePowerDist(ctx context.Context) {
+	params := k.GetParams(ctx)
+	btcTipHeight, err := k.GetCurrentBTCHeight(ctx)
+	if err != nil {
+		panic(err) // only possible upon programming error
+	}
+	wValue := k.btccKeeper.GetParams(ctx).CheckpointFinalizationTimeout
+
+	// prepare metrics for {active, inactive} finality providers,
+	// {pending, active, unbonded BTC delegations}, and total staked Bitcoins
+	// NOTE: slashed finality providers and BTC delegations are recorded upon
+	// slashing events rather than here
+	var (
+		numFPs        int    = 0
+		numStakedSats uint64 = 0
+		numDelsMap           = map[types.BTCDelegationStatus]int{
+			types.BTCDelegationStatus_PENDING:  0,
+			types.BTCDelegationStatus_ACTIVE:   0,
+			types.BTCDelegationStatus_UNBONDED: 0,
+		}
+	)
+	// for voting power and rewards
+	dc := types.NewVotingPowerDistCache()
+
+	// iterate over all finality providers to find out non-slashed ones that have
+	// positive voting power
+	k.IterateActiveFPs(
+		ctx,
+		func(fp *types.FinalityProvider) bool {
+			fpDistInfo := types.NewFinalityProviderDistInfo(fp)
+
+			// iterate over all BTC delegations under the finality provider
+			// in order to accumulate voting power dist info for it
+			k.IterateBTCDelegations(ctx, fp.BtcPk, func(btcDel *types.BTCDelegation) bool {
+				// accumulate voting power and reward distribution cache
+				fpDistInfo.AddBTCDel(btcDel, btcTipHeight, wValue, params.CovenantQuorum)
+
+				// record metrics
+				numStakedSats += btcDel.VotingPower(btcTipHeight, wValue, params.CovenantQuorum)
+				numDelsMap[btcDel.GetStatus(btcTipHeight, wValue, params.CovenantQuorum)]++
+
+				return true
+			})
+
+			if fpDistInfo.TotalVotingPower > 0 {
+				dc.AddFinalityProviderDistInfo(fpDistInfo)
+			}
+
+			return true
+		},
+	)
+	// record metrics for finality providers and total staked BTCs
+	numActiveFPs := min(numFPs, int(params.MaxActiveFinalityProviders))
+	types.RecordActiveFinalityProviders(numActiveFPs)
+	types.RecordInactiveFinalityProviders(numFPs - numActiveFPs)
+	numStakedBTCs := float32(numStakedSats / SatoshisPerBTC)
+	types.RecordMetricsKeyStakedBitcoins(numStakedBTCs)
+	// record metrics for BTC delegations
+	for status, num := range numDelsMap {
+		types.RecordBTCDelegations(num, status)
+	}
+
+	// filter out top `MaxActiveFinalityProviders` active finality providers in terms of voting power
+	maxNumActiveFPs := k.GetParams(ctx).MaxActiveFinalityProviders
```

Comment:
```suggestion
	paramMaxNumActiveFps := params.MaxActiveFinalityProviders
```
Also, this line can be inserted before Line 68